### DISCUSSION TITLE
Is it possible to get ContainerRequestContext or HttpHeaders in Jackson serializers/deserializers?

### DISCUSSION THREAD

**Discussion by marcushelsing** [FasterXML/jackson-databind#4431] [General]
I have a problem with not knowing the ContainerRequestContext or HttpHeaders content i my jackson serialisers and deserialisers. I have implemented a solution with ImprovedThreadLocal but this is not a stable solution. Can anyone help me here?

**Comment by yawkat** [comment]
that depends more on your framework. You can pass in context objects through the withAttribute api for the deserializer to read.

**Reply by marcushelsing** [reply]
Ok, so can I pass ContainerRequestContext through DeserializationContext object? Is that threadsafe?

**Reply by yawkat** [reply]
the idea is that you create an ObjectReader, and then do `ObjectReader.withAttribute(...).read(...)` in every request. since withAttribute creates a new ObjectReader, different requests cannot interfere with each other.

**Reply by marcushelsing** [reply]
Ok, thanks for your quick resonse. 
I have my custom ContainerRequestFilter who has access to ContainerRequestContext and in filter method Ill then create the ObjectReader instance containing ContainerRequestContext.  Can you give me som code examples, please?

**Reply by marcushelsing** [reply]
We use Jersey/Jackson, no Spring

**Reply by marcushelsing** [reply]
Is this a possible way forward? 
```
public class MyValidatorFilter implements ContainerRequestFilter {

    @Override
    public void filter(ContainerRequestContext requestContext) throws IOException {
        ObjectMapper          mapper                = new ObjectMapper();
        DeserializationConfig deserializationConfig = mapper.getDeserializationConfig();
        DeserializationConfig modifiedConfig        = deserializationConfig.withAttribute("datastring", requestContext.getProperty("datastring"));
        mapper.setConfig(modifiedConfig);
        
        ...
    }
}
public class MyDeserializer<T> extends StdDeserializer<T> {
    @Override
    public T deserialize(JsonParser pJsonParser, DeserializationContext pDeserializationContext)
    throws IOException {

        String token = (String) pDeserializationContext.getAttribute("datastring");

        return null;
    }
}
```

**Reply by cowtowncoder** [reply]
Not quite: you cannot (re)configure `ObjectMapper` in thread-safe manner. And you definitely do not want to be creating new ObjectMappers for every request (horrible performance).

But what you can do is to create `ObjectReader` from `ObjectMapper` and attach attributes on per-request `ObjectReader` instance.
